### ISSUE TITLE
bullet-filter-algorithm ignored by ray tests

### ISSUE THREAD

**Issue by wezu** [panda3d/panda3d#302]
When setting `bullet-filter-algorithm` to `groups-mask` ray tests ignore the group collisions flags and just check if the ray and any collision solids have any common bits.
Test code: https://gist.github.com/wezu/fd987811d29095b558b06d1f28ce166c

**Comment by ArsThaumaturgis** [comment]
Ah, after some experimentation, I think that I see what's happening:

The mask used by the ray-cast doesn't operate as a group-mask. Instead, it's asking which groups it should collide with.

For example, if you change your code there such that the ray's mask uses bit 1, you should find that you can now hit the box. If you set it to use bit 0, you should now be able to hit the floor. Setting both bits allows you to hit both. Since it's currently set to 3, and nothing is in group 3, it doesn't hit anything.

Whether this is what's desired is a question. It's perhaps not intuitive under the groups-mask system. On the other hand, one might argue that a once-off ray-cast is a little different to normal physics interactions: one generally doesn't have to worry about rays interacting amongst themselves, or physical consequences of rays hitting objects.

**Comment by wezu** [comment]
I'm willing to agree it may be a 'feature'. You can set a mask on a ray that will collide with exactly what's needed but you need to keep track of two different systems. 

Say you have node_a with mask 1 (mask.set_bit(1)) and node_b with mask 2 (mask.set_bit(2)), after calling world.set_group_collision_flag(1,2, True) node_a will collide with node_b but a ray using any of these masks will only collide with either node_a or node_b.  

If this is indeed by design then it needs to be mentioned in the manual (and reference).

**Comment by ArsThaumaturgis** [comment]
Indeed, if it remains this way then I think that having the manual explicitly describe how it works would be a good idea.

**Comment by ArsThaumaturgis** [comment]
Suggested manual text:

The ray-test methods are a little different. They are tests, not persistent objects, and are not placed into groups. Instead, you indicate which groups you want to test against. Note that this ignores any group-collision flags: all groups specified in the ray-test mask are tested against.

**Comment by rdb** [comment]
Indeed, ray tests don't use the "normal" filter algorithm; you directly specify which groups you want to test against.

I'm not sure how others feel about this design but I'm personally inclined to say that the existing behaviour makes sense; as Thaumaturge points out, a ray test is really a kind of "direct" query operation and not anything like a physics body, so it makes sense that you are directly specifying which groups you want to query against. Furthermore, you already have maximal control over the possible interactions you can make using this mask; there's no extra control gained from using the group collision matrix. It may even be a nuisance for some to (potentially) have to set up a whole new group for their ray tests.

However, for those expecting it to work as wezu did, we could add a convenience method to BulletWorld that remaps a group mask into a mask for use in a query using the group collision matrix.

I am also fine with adding this to the documentation.
